src/System.IO.Pipelines/PipeCompletion.cs

```diff
@@ -1,58 +1,121 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System.Threading;
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
 
 namespace System.IO.Pipelines
 {
     internal struct PipeCompletion
     {
+        private static readonly ArrayPool<PipeCompletionCallback> CompletionCallbackPool = ArrayPool<PipeCompletionCallback>.Shared;
+
+        private const int InitialCallbacksSize = 1;
         private static readonly Exception _completedNoException = new Exception();
 
-        private Exception _exception;
 #if COMPLETION_LOCATION_TRACKING
         private string _completionLocation;
 #endif
+        public Exception Exception { get; set; }
+
+        private PipeCompletionCallback[] _callbacks;
+        private int _callbackCount;
+
+        public string Location
+        {
+            get
+            {
+#if COMPLETION_LOCATION_TRACKING
+                return _completionLocation;
+#else
+                return null;
+#endif
+            }
+        }
 
-        public bool IsCompleted => _exception != null;
+        public bool IsCompleted => Exception != null;
 
-        public void TryComplete(Exception exception = null)
+        public bool TryComplete(Exception exception = null)
         {
 #if COMPLETION_LOCATION_TRACKING
             _completionLocation = Environment.StackTrace;
 #endif
-            // Set the exception object to the exception passed in or a sentinel value
-            Interlocked.CompareExchange(ref _exception, exception ?? _completedNoException, null);
+            if (Exception == null)
+            {
+                // Set the exception object to the exception passed in or a sentinel value
+                Exception = exception ?? _completedNoException;
+            }
+            return _callbackCount > 0;
```

Comment:
What's the benefit of checking the _callbackCount inside the lock? It seems to me this could be checked outside the lock if its only purpose is to avoid unnecessarily scheduling Invoke[Reader|Writer]CompletionCallbacks.